### ISSUE TITLE
MAE loss doesn't decrease?

### ISSUE THREAD

**Issue by wa1618i** [ChenglongChen/tensorflow-DeepFM#1]
Not sure why but if I use deepFM for regression problem, MAE not improving?
Example:

import numpy as np
from DeepFM import DeepFM
import tensorflow as tf
import pandas as pd
from sklearn.cross_validation import KFold
from sklearn.metrics import mean_absolute_error
from DataReader import FeatureDictionary, DataParser
train=pd.DataFrame(np.random.randn(100, 100).astype('float'))
y=pd.DataFrame(np.random.randn(100,1).astype('float'))
train['loss']=y
test=pd.DataFrame(np.random.randn(100, 100).astype('float'))
cols=[c for c in train.columns if 'loss' not in c]
fd = FeatureDictionary(dfTrain=train, dfTest=test,numeric_cols=cols,ignore_cols='loss')
data_parser = DataParser(feat_dict=fd)
Xi_train, Xv_train,y_train = data_parser.parse(df=train, has_label=True)
Xi_test, Xv_test = data_parser.parse(df=test)
dfm_params = {
    "use_fm": True,
    "use_deep": True,
    "embedding_size": 8,
    "dropout_fm": [1.0, 1.0],
    "deep_layers": [32, 32],
    "dropout_deep": [0.5, 0.5, 0.5],
    "deep_layers_activation": tf.nn.relu,
    "epoch": 30,
    "batch_size": 1024,
    "learning_rate": 0.001,
    "optimizer_type": "adam",
    "batch_norm": 1,
    "batch_norm_decay": 0.995,
    "l2_reg": 0.01,
    "verbose": True,
    "metric": mean_absolute_error,
    "random_seed": 42,'greater_is_better':False
}
dfm_params["feature_size"] = fd.feat_dim
dfm_params["field_size"] = len(Xi_train[0])
_get = lambda x, l: [x[i] for i in l]
del train['loss']
train=train.values
y=y.values
y=y.reshape((-1))
folds = KFold(len(y),8, shuffle=True, random_state=2016)
for i, (train_idx, valid_idx) in enumerate(folds):
        Xi_train_, Xv_train_, y_train_ = _get(Xi_train, train_idx), _get(Xv_train, train_idx), _get(y_train, train_idx)
        Xi_valid_, Xv_valid_, y_valid_ = _get(Xi_train, valid_idx), _get(Xv_train, valid_idx), _get(y_train, valid_idx)
dfm = DeepFM(**dfm_params)
dfm.fit(Xi_train_, Xv_train_, y_train_, Xi_valid_, Xv_valid_, y_valid_)

[1] train-result=0.9263, valid-result=0.9745 [0.0 s]
[2] train-result=0.9263, valid-result=0.9745 [0.0 s]
[3] train-result=0.9263, valid-result=0.9745 [0.0 s]
[4] train-result=0.9263, valid-result=0.9745 [0.0 s]
[5] train-result=0.9263, valid-result=0.9745 [0.0 s]
[6] train-result=0.9263, valid-result=0.9745 [0.0 s]
[7] train-result=0.9263, valid-result=0.9746 [0.0 s]
[8] train-result=0.9263, valid-result=0.9746 [0.0 s]
[9] train-result=0.9263, valid-result=0.9746 [0.0 s]
[10] train-result=0.9263, valid-result=0.9746 [0.0 s]
[11] train-result=0.9264, valid-result=0.9746 [0.0 s]
[12] train-result=0.9264, valid-result=0.9746 [0.0 s]
[13] train-result=0.9264, valid-result=0.9746 [0.0 s]
[14] train-result=0.9264, valid-result=0.9746 [0.0 s]
[15] train-result=0.9264, valid-result=0.9746 [0.0 s]
[16] train-result=0.9264, valid-result=0.9747 [0.0 s]
[17] train-result=0.9264, valid-result=0.9747 [0.0 s]
[18] train-result=0.9264, valid-result=0.9747 [0.0 s]
[19] train-result=0.9264, valid-result=0.9747 [0.0 s]
[20] train-result=0.9264, valid-result=0.9747 [0.0 s]
[21] train-result=0.9264, valid-result=0.9747 [0.0 s]
[22] train-result=0.9264, valid-result=0.9747 [0.0 s]
[23] train-result=0.9264, valid-result=0.9747 [0.0 s]
[24] train-result=0.9264, valid-result=0.9747 [0.0 s]
[25] train-result=0.9264, valid-result=0.9747 [0.0 s]
[26] train-result=0.9264, valid-result=0.9747 [0.0 s]
[27] train-result=0.9264, valid-result=0.9748 [0.0 s]
[28] train-result=0.9264, valid-result=0.9748 [0.0 s]
[29] train-result=0.9264, valid-result=0.9748 [0.0 s]
[30] train-result=0.9264, valid-result=0.9748 [0.0 s]

**Comment by ChenglongChen** [comment]
Hi,

The original implementation only supports classification task (as in CTR prediction). I have added support for regression. Please clone the new version. To use it, you can set loss_type as mse. Accordingly, you should use eval_metric for regression, e.g., mse or mae. 

To following code snippet trains a DeepFM model for regression for the sake of example. You can have a try.

```
import numpy as np
from DeepFM import DeepFM
import tensorflow as tf
import pandas as pd
from sklearn.cross_validation import KFold
from sklearn.metrics import mean_absolute_error
train=pd.DataFrame(np.random.randn(10000, 100).astype('float'))
y=pd.DataFrame(np.random.randn(10000,1).astype('float'))
train['target']=y
train['id']=np.arange(train.shape[0])
test=pd.DataFrame(np.random.randn(10000, 100).astype('float'))
test['id']=np.arange(test.shape[0])
cols=[c for c in train.columns if 'target' != c]
fd = FeatureDictionary(dfTrain=train, dfTest=test,numeric_cols=cols,ignore_cols=['target'])
data_parser = DataParser(feat_dict=fd)
Xi_train, Xv_train,y_train = data_parser.parse(df=train, has_label=True)
Xi_test, Xv_test, ids_test = data_parser.parse(df=test)
dfm_params = {
    "use_fm": True,
    "use_deep": True,
    "embedding_size": 8,
    "dropout_fm": [1.0, 1.0],
    "deep_layers": [32, 32],
    "dropout_deep": [0.5, 0.5, 0.5],
    "deep_layers_activation": tf.nn.relu,
    "epoch": 30,
    "batch_size": 100,
    "learning_rate": 0.001,
    "optimizer_type": "adam",
    "batch_norm": 1,
    "batch_norm_decay": 0.995,
    "l2_reg": 0.01,
    "verbose": True,
    "loss_type": "mse",
    "eval_metric": mean_absolute_error,
    "random_seed": 42,
    "greater_is_better":False
}
dfm_params["feature_size"] = fd.feat_dim
dfm_params["field_size"] = len(Xi_train[0])
get = lambda x, l: [x[i] for i in l]
y=y.values
y=y.reshape((-1))
folds = KFold(len(y),8, shuffle=True, random_state=2016)
for i, (train_idx, valid_idx) in enumerate(folds):
    Xi_train_, Xv_train_, y_train_ = get(Xi_train, train_idx), get(Xv_train, train_idx), get(y_train, train_idx)
    Xi_valid_, Xv_valid_, y_valid_ = get(Xi_train, valid_idx), get(Xv_train, valid_idx), get(y_train, valid_idx)
    dfm = DeepFM(**dfm_params)
    dfm.fit(Xi_train, Xv_train, y_train, Xi_valid_, Xv_valid_, y_valid_)
```

To output log:
```
[1] train-result=0.9090, valid-result=0.8740 [1.9 s]
[2] train-result=0.8618, valid-result=0.8240 [1.8 s]
[3] train-result=0.8331, valid-result=0.7971 [1.7 s]
[4] train-result=0.8191, valid-result=0.7836 [1.5 s]
[5] train-result=0.8104, valid-result=0.7770 [1.6 s]
[6] train-result=0.8078, valid-result=0.7756 [1.8 s]
[7] train-result=0.8055, valid-result=0.7751 [1.7 s]
[8] train-result=0.8050, valid-result=0.7749 [1.8 s]
[9] train-result=0.8047, valid-result=0.7738 [1.6 s]
[10] train-result=0.8048, valid-result=0.7740 [1.8 s]
[11] train-result=0.8039, valid-result=0.7736 [1.7 s]
[12] train-result=0.8040, valid-result=0.7734 [1.8 s]
[13] train-result=0.8038, valid-result=0.7735 [1.6 s]
[14] train-result=0.8029, valid-result=0.7717 [1.9 s]
[15] train-result=0.8027, valid-result=0.7716 [2.0 s]
[16] train-result=0.8019, valid-result=0.7712 [1.9 s]
[17] train-result=0.8000, valid-result=0.7696 [1.6 s]
[18] train-result=0.7985, valid-result=0.7678 [2.3 s]
[19] train-result=0.7957, valid-result=0.7651 [1.9 s]
[20] train-result=0.7929, valid-result=0.7602 [1.9 s]
[21] train-result=0.7900, valid-result=0.7576 [2.0 s]
[22] train-result=0.7874, valid-result=0.7542 [2.2 s]
[23] train-result=0.7839, valid-result=0.7505 [2.1 s]
[24] train-result=0.7814, valid-result=0.7472 [1.8 s]
[25] train-result=0.7789, valid-result=0.7449 [2.0 s]
[26] train-result=0.7765, valid-result=0.7423 [1.7 s]
[27] train-result=0.7750, valid-result=0.7401 [1.6 s]
[28] train-result=0.7735, valid-result=0.7405 [1.8 s]
[29] train-result=0.7722, valid-result=0.7394 [1.6 s]
[30] train-result=0.7704, valid-result=0.7393 [1.6 s]
```

Hope it helps.
